cocotb/drivers/amba.py

```diff
@@ -60,153 +89,401 @@ def __init__(self, entity, name, clock, **kwargs):
+            size: The size (in bytes) of each beat. Defaults to None (width of
+                the data bus).
+            burst: The burst type, either ``FIXED``, ``INCR`` or ``WRAP``.
+                Defaults to ``INCR``.
+            byte_enable: Which bytes in value to actually write. Defaults to
+                None (write all bytes).
+            address_latency: Delay before setting the address (in clock
+                cycles). Default is no delay.
+            data_latency: Delay before setting the data value (in clock
+                cycles).
                 Default is no delay.
             sync: Wait for rising edge on clock initially.
                 Defaults to True.
 
-        Returns:
-            The write response value.
-
         Raises:
+            ValueError: If any of the input parameters is invalid.
             AXIProtocolError: If write response from AXI is not ``OKAY``.
         """
-        if sync:
-            await RisingEdge(self.clock)
 
-        c_addr = cocotb.fork(self._send_write_address(address,
-                                                      delay=address_latency))
-        c_data = cocotb.fork(self._send_write_data(value,
-                                                   byte_enable=byte_enable,
-                                                   delay=data_latency))
+        if not isinstance(value, collections.abc.Sequence):
+            value = (value,)    # If value is not a sequence, make it
 
-        if c_addr:
-            await c_addr.join()
-        if c_data:
-            await c_data.join()
+        if not isinstance(byte_enable, collections.abc.Sequence):
+            byte_enable = (byte_enable,)    # Same for byte_enable
 
-        # Wait for the response
-        while True:
-            await ReadOnly()
-            if self.bus.BVALID.value and self.bus.BREADY.value:
-                result = self.bus.BRESP.value
-                break
-            await RisingEdge(self.clock)
+        if size is None:
+            size = len(self.bus.WDATA) // 8
+        else:
+            AXI4Master._check_size(size, len(self.bus.WDATA) // 8)
 
-        await RisingEdge(self.clock)
+        AXI4Master._check_length(len(value), burst)
+        AXI4Master._check_4kB_boundary_crossing(address, burst, size,
+                                                len(value))
 
-        if int(result):
-            raise AXIProtocolError("Write to address 0x%08x failed with BRESP: %d"
-                                   % (address, int(result)))
+        write_address = self._send_write_address(address, len(value), burst,
+                                                 size, address_latency, sync)
 
-        return result
+        write_data = self._send_write_data(value, data_latency, byte_enable,
+                                           sync)
+
+        await Combine(cocotb.fork(write_address), cocotb.fork(write_data))
+
+        async with self.write_response_busy:
```

Comment:
Can write responses can come back immediately? If you waiting until after the address and data are written out to start checking the response channel may lead to the response being missed?

Reply:
No, that's forbidden by the standard.

A3.3.1: "The slave must wait for AWVALID, AWREADY, WVALID, and WREADY to be asserted before asserting BVALID. The slave must also wait for WLAST to be asserted before asserting BVALID because the write response, BRESP must be signaled only after the last data transfer of a write transaction"